Wikipedia/Code/WMFAppViewController.m

```diff
@@ -370,6 +410,88 @@ - (WMFDailyStatsLoggingFunnel *)statsFunnel {
     return _statsFunnel;
 }
 
+#pragma mark - Content Sources
+
+- (void)preloadContentSourcesIfNeededWithCompletion:(void (^)(void))completion {
+    
+    if([[NSUserDefaults wmf_userDefaults] wmf_didMigrateToNewFeed]){
+        if(completion){
+            completion();
+        }
+    }else{
+        WMFTaskGroup* group = [WMFTaskGroup new];
+        [self.contentSources enumerateObjectsUsingBlock:^(id<WMFContentSource>  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+            
+            if([obj respondsToSelector:@selector(preloadContentForNumberOfDays:completion:)]){
+                [group enter];
+                [obj preloadContentForNumberOfDays:2 completion:^{
+                    [group leave];
+                }];
+            }
+        }];
+        
+        [group waitInBackgroundWithCompletion:^{
+            [[NSUserDefaults wmf_userDefaults] wmf_setDidMigrateToNewFeed:YES];
+            if(completion){
+                completion();
+            }
+        }];
+    }
+
+}
+
+
+- (void)updateFeedSourcesWithCompletion:(dispatch_block_t)completion{
+    WMFTaskGroup* group = [WMFTaskGroup new];
+    [self.contentSources enumerateObjectsUsingBlock:^(id<WMFContentSource>  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+        [group enter];
+        [obj loadNewContentForce:NO completion:^{
+            [group leave];
+        }];
+    }];
+    
+    //TODO: nearby doesnt always fire.
+    //May need to time it out or exclude
+    [group waitInBackgroundWithCompletion:^{
+        if(completion){
+            completion();
+        }
+    }];
+}
+
+- (void)startContentSources{
+    [self.contentSources makeObjectsPerformSelector:@selector(startUpdating)];
+}
+
+- (void)stopContentSources{
+    [self.contentSources makeObjectsPerformSelector:@selector(stopUpdating)];
+}
+
+- (WMFNearbyContentSource*)nearbyContentSource{
+    return [self.contentSources bk_match:^BOOL(id<WMFContentSource> obj) {
+        return [obj isKindOfClass:[WMFNearbyContentSource class]];
+    }];
+}
+
+- (NSArray<id<WMFContentSource>>*)contentSources{
+    NSParameterAssert(self.contentStore);
```

Comment:
This assert is tripping when I try to remove a language from the languages welcome screen.